### DISCUSSION TITLE
Why doesn't KeystoneJS support autoincrement on MySQL?

### DISCUSSION THREAD

**Discussion by jinta4020** [keystonejs/keystone#8329] [Questions] [answered]
The KeystoneJS documentation suggests that when using an Integer field with the `defaultValue: { kind: 'autoincrement' }` option on MySQL, the field should also be indexed using `isIndexed: true` or `isIndexed: 'unique'`. However, the documentation for the `bigint` field explicitly states that `{ kind: 'autoincrement' }` is only supported on PostgreSQL.

https://keystonejs.com/docs/guides/choosing-a-database#auto-increment-integer-fields
https://keystonejs.com/docs/fields/bigint

I attempted to use `defaultValue: { kind: 'autoincrement' }` on my app that is using MySQL, but it gave me an error saying that `this is not supported on mysql`.

I'm now wondering why KeystoneJS doesn't support auto-increment on MySQL, and whether the documentation on choosing a database and the `bigint` field contains mistakes. Can anyone provide clarification on this issue?

**Comment by borisno2** [comment, marked as answer]
Hi @jinta4020 
The choosing a database page was specifically written with the `integer` field in mind (https://keystonejs.com/docs/fields/integer), not the `BigInt` field. In Prisma a `BigInt` in autoincrement on MySQL is a different type(`@db.UnsignedBigInt @default(autoincrement())` while on Postgres adding an `autoincrement` does not change the Native database type attribute in the Prisma Schema (see: https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#bigint). Currently, the `BigInt` field is not set up to change its native type based on the `defaultValue` and instead errors.

**Reply by jinta4020** [reply]
I believe that this issue is not related to the `BigInt` field type. The `Integer` field documentation (https://keystonejs.com/docs/fields/integer) includes the same phrase: `{ kind: 'autoincrement' } is only supported on PostgreSQL.`

I have successfully configured the autoincrement feature in KeystoneJS using the MySQL database provider for the `idField`, and it generated the correct schema.

```schema.ts
import { list } from '@keystone-6/core'
import { allOperations, allowAll } from '@keystone-6/core/access'
import { text, integer, bigInt } from '@keystone-6/core/fields'
import type { Lists } from '.keystone/types'

export const lists: Lists = {
  Autoincrement: list({
    access: ...,
    fields: ...,
    db: {
      idField: {
        kind: 'autoincrement',
        type: 'BigInt', // or type: 'Int',
      },
    },
  }),
}
```

```prisma.schema
id         BigInt @id @default(autoincrement())
// or
id         Int    @id @default(autoincrement())
```

However, I encountered an error when I tried to use the autoincrement feature on a normal field. The error message says `Error: The integer field at Autoincrement.test specifies defaultValue: { kind: 'autoincrement' }, this is not supported on mysql.`

Here's the code that produced the error:

```schema.ts
export const lists: Lists = {
  Autoincrement: list({
    access: ...,
    fields: {
      test: integer({ // or bigInt
        defaultValue: { kind: 'autoincrement' },
        isIndexed: 'unique',
      }),
    },
  }),
}
```

In native prisma, I can use `autoincrement` like this.

```prisma.schema
test         BigInt @default(autoincrement())
// or
test         Int    @default(autoincrement())
```

I expected the autoincrement feature to work with the `Integer` field at least.

The behavior seems odd to me, and if there's no reason for it, I would like to submit a pull request to fix it. Thank you.

**Reply by borisno2** [reply]
Hey @jinta4020 Thanks for your reply, there is no reason that I am aware of as to why these paths have not been implemented/allowed. That would be great if you could submit a PR to resolve this! Thank you!
